src/app/util/privilege-storage.cpp

```diff
@@ -0,0 +1,142 @@
+/**
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "privilege-storage.h"
+
+#include <zap-generated/access.h>
+
+#include <lib/support/CodeUtils.h>
+
+#include <cstdint>
+
+using chip::AttributeId;
+using chip::ClusterId;
+using chip::CommandId;
+using chip::EventId;
+
+namespace {
+
+static_assert(GENERATED_ACCESS_PRIVILEGE__VIEW == kMatterAccessPrivilegeView,
+              "Generated privilege value must match privilege value in code");
+static_assert(GENERATED_ACCESS_PRIVILEGE__OPERATE == kMatterAccessPrivilegeOperate,
+              "Generated privilege value must match privilege value in code");
+static_assert(GENERATED_ACCESS_PRIVILEGE__MANAGE == kMatterAccessPrivilegeManage,
+              "Generated privilege value must match privilege value in code");
+static_assert(GENERATED_ACCESS_PRIVILEGE__ADMINISTER == kMatterAccessPrivilegeAdminister,
+              "Generated privilege value must match privilege value in code");
+
+#ifdef GENERATED_ACCESS_READ_ATTRIBUTE__CLUSTER
+namespace GeneratedAccessReadAttribute {
+constexpr ClusterId kCluster[]     = GENERATED_ACCESS_READ_ATTRIBUTE__CLUSTER;
+constexpr AttributeId kAttribute[] = GENERATED_ACCESS_READ_ATTRIBUTE__ATTRIBUTE;
+constexpr uint8_t kPrivilege[]     = GENERATED_ACCESS_READ_ATTRIBUTE__PRIVILEGE;
+static_assert(ArraySize(kCluster) == ArraySize(kAttribute) && ArraySize(kAttribute) == ArraySize(kPrivilege),
+              "Generated parallel arrays must be same size");
+} // namespace GeneratedAccessReadAttribute
+#endif
+
+#ifdef GENERATED_ACCESS_WRITE_ATTRIBUTE__CLUSTER
+namespace GeneratedAccessWriteAttribute {
+constexpr ClusterId kCluster[]     = GENERATED_ACCESS_READ_ATTRIBUTE__CLUSTER;
```

Comment:
Readability-wise, I would imagine a single array of pairs of cluster/attribute/privilege would be clearer. However I can see how that causes alignment issues since privilege is small and we don't want it padded.

So if this split was intentional (and using packed is not practical) it would be nice to document the choice and reasoning.

Reply:
The comments about "parallel arrays" should now make it clear they are related and give a term to search for if there are questions why they are used.